### PR TITLE
ENH: Dropout in transformer layers in DeepMIL

### PR DESCRIPTION
**Pull request by harshita-s**
In this PR:
- `transformer_dropout` parameter is added to `TransformerPooling` and `TransformerPoolingBenchmark` pooling layers.
- Tests are updated with the `transformer_dropout` parameter.

### CODE REVIEW

File: hi-ml/testhiml/testhiml/test_attentionlayers.py [hunk 1]
```diff
@@ -23,6 +23,8 @@ def _test_attention_layer(attentionlayer: nn.Module, dim_in: int, dim_att: int,
         pass
     elif isinstance(attentionlayer, MaxPoolingLayer):
         pass
+    elif isinstance(attentionlayer, TransformerPoolingBenchmark):
```
Comment (dccastro) [inline hi-ml/testhiml/testhiml/test_attentionlayers.py:26]: [Minor] To avoid having too many cases here, we can instead use a single `isinstance()` with a tuple of all the types we want to check at once:
```python
if not isinstance(attentionlayer, (Class1, Class2, ...)):
    pooled_features = attn_weights @ features.flatten(start_dim=1)
    assert allclose(pooled_features, output_features)
```

File: hi-ml/testhiml/testhiml/test_attentionlayers.py [hunk 2]
```diff
@@ -59,8 +61,18 @@ def test_max_pooling(dim_in: int, batch_size: int,) -> None:
 @pytest.mark.parametrize("num_heads", [1, 2])
 @pytest.mark.parametrize("dim_in", [4, 8])   # dim_in % num_heads must be 0
 @pytest.mark.parametrize("batch_size", [1, 7])
-def test_transformer_pooling(num_layers: int, num_heads: int, dim_in: int, batch_size: int) -> None:
+@pytest.mark.parametrize("transformer_dropout", [0.0, 0.3])
```
Comment (dccastro) [inline hi-ml/testhiml/testhiml/test_attentionlayers.py:64]: We now have 64 combinations of inputs for this test. Do you think it might make sense to test the dropout options in a separate test function, keeping this one focused on compatibility of tensor shapes?

File: hi-ml/testhiml/testhiml/test_attentionlayers.py [hunk 3]
```diff
@@ -59,8 +61,18 @@ def test_max_pooling(dim_in: int, batch_size: int,) -> None:
...
+@pytest.mark.parametrize("dim_hid", [1, 4])
+def test_transformer_pooling(num_layers: int, num_heads: int, dim_in: int,
+                             batch_size: int, dim_hid: int, transformer_dropout: float) -> None:
     transformer_pooling = TransformerPooling(num_layers=num_layers,
                                              num_heads=num_heads,
-                                             dim_representation=dim_in).eval()
+                                             dim_representation=dim_in,
+                                             transformer_dropout=transformer_dropout).eval()
     _test_attention_layer(transformer_pooling, dim_in=dim_in, dim_att=1, batch_size=batch_size)
+    transformer_pooling_benchmark = TransformerPoolingBenchmark(num_layers=num_layers,
```
Comment (dccastro) [inline hi-ml/testhiml/testhiml/test_attentionlayers.py:73]: I'd suggest testing this component in a separate test function or test case, so we can catch errors separately for both. Also, they have different input arguments.

**Review by dccastro — APPROVED** [review]

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py [hunk 4]
```diff
@@ -150,13 +155,18 @@ def get_pooling_layer(self, num_encoding: int) -> Tuple[nn.Module, int]:
...
             self.pool_out_dim = 1  # currently this is hardcoded in forward of the TransformerPooling
         elif self.pool_type == TransformerPoolingBenchmark.__name__:
             pooling_layer = TransformerPoolingBenchmark(
-                self.num_transformer_pool_layers, self.num_transformer_pool_heads, num_encoding, self.pool_hidden_dim
-            )
+                self.num_transformer_pool_layers,
+                self.num_transformer_pool_heads,
+                num_encoding,
+                self.pool_hidden_dim,
+                self.transformer_dropout)
```
[the remark this answers is not in the document]
Comment (harshita-s) [reply hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py]: Sure! Now passing keyword arguments

File: hi-ml/testhiml/testhiml/test_attentionlayers.py (hunk 2 above)
Comment (ant0nsc) [reply hi-ml/testhiml/testhiml/test_attentionlayers.py:64]: +1 on avoiding combinatorial explosion

File: hi-ml/testhiml/testhiml/test_attentionlayers.py (hunk 3 above)
Comment (harshita-s) [reply hi-ml/testhiml/testhiml/test_attentionlayers.py:76]: Thanks, now separated the tests

File: hi-ml/testhiml/testhiml/test_attentionlayers.py (hunk 2 above)
Comment (harshita-s) [reply hi-ml/testhiml/testhiml/test_attentionlayers.py]: We currently have one test function for each type of pooling layer, so I don't think we should have another function only `transformer_dropout` which is being used in 2/6 layers. To prevent too many combinations, I have fixed `transformer_dropout` parameter to one value for now. I think we could test it without parametrizing. What do you think?

File: hi-ml/testhiml/testhiml/test_attentionlayers.py (hunk 1 above)
Comment (harshita-s) [reply hi-ml/testhiml/testhiml/test_attentionlayers.py]: Thanks, done

File: hi-ml/testhiml/testhiml/test_attentionlayers.py (hunk 2 above)
Comment (dccastro) [reply hi-ml/testhiml/testhiml/test_attentionlayers.py]: It should be fine to hard-code the dropout rate, as we don't expect any substantial behavioural changes.
